feat(treesitter): add optional modeline arguments to query.set

tadmccorkle opened the pull request:
# Problem

`vim.treesitter.query.set` only supports overriding queries defined in runtime files. It does not provide a way to extend existing queries. There is not an easy way to customize query behavior based on dynamic configuration (e.g., as a plugin).

# Solution

Add optional arguments to `vim.treesitter.query.set` that correspond to the treesitter query modelines supported in runtime query files.

## Update 2024-02-23

- Support added for query overrides from `vim.treesitter.query.set` to be
  applied for an entire language or for individual buffers by providing
  and optional `bufnr` argument to both `vim.treesitter.query.set` and
  `vim.treesitter.query.get`.

Implementation notes:

- A given query can be extended by multiple callers. For example,
  multiple plugins can provide extensions to the same query. As such,
  parsing of query overrides and extensions has been moved to
  `vim.treesitter.query.get`.
- Queries can be extended (and also now overridden) for an entire
  language or for individual buffers. Buffers with buffer-only overrides
  or extensions will also include any global language extensions.
- Tree-sitter fold, highlighter, and languagetree modules have been
  updated to support buffer-only query overrides and extensions.
- Memoize cache invalidation was necessary to implement this feature
  with consistent behavior (i.e., seeing query changes applied via
  `vim.treesitter.query.set` immediately). The
  `vim.treesitter.query.get` memoize hash is now cleared when calling
  `vim.treesitter.query.set`. When setting buffer-only queries the
  corresponding function call is removed from the memoize cache. When
  setting global lang queries the entire memoize cache is cleared. This
  scheme was picked under the assumption that global lang queries will
  typically be set before getting queries, whereas buffer-only queries
  may be set and changed throughout a session.

Closes #26965.

tadmccorkle committed:
feat(treesitter): add optional modeline arguments to query.set

Problem:

- 'vim.treesitter.query.set' only supports overriding queries defined in
  runtime files. It does not provide a way to extend existing queries.
  There is not an easy way to customize query behavior based on dynamic
  configuration (e.g., as a plugin).

Solution:

- Add optional arguments to 'vim.treesitter.query.set' that correspond to
  to the treesitter query modelines supported in runtime query files.
- Add support for query overrides from `vim.treesitter.query.set` to be
  applied for an entire language or for individual buffers by providing
  and optional `bufnr` argument to both `vim.treesitter.query.set` and
  `vim.treesitter.query.get`.

Implementation notes:

- A given query can be extended by multiple callers. For example,
  multiple plugins can provide extensions to the same query. As such,
  parsing of query overrides and extensions has been moved to
  `vim.treesitter.query.get`.
- Queries can be extended (and also now overridden) for an entire
  language or for individual buffers. Buffers with buffer-only overrides
  or extensions will also include any global language extensions.
- Tree-sitter fold, highlighter, and languagetree modules have been
  updated to support buffer-only query overrides and extensions.
- Memoize cache invalidation was necessary to implement this feature
  with consistent behavior (i.e., seeing query changes applied via
  `vim.treesitter.query.set` immediately). The
  `vim.treesitter.query.get` memoize hash is now cleared when calling
  `vim.treesitter.query.set`. When setting buffer-only queries the
  corresponding function call is removed from the memoize cache. When
  setting global lang queries the entire memoize cache is cleared. This
  scheme was picked under the assumption that global lang queries will
  typically be set before getting queries, whereas buffer-only queries
  may be set and changed throughout a session.

tadmccorkle commented:
Regarding tests, I was having trouble coming up with a way to test inheriting a query with the default Neovim runtime files. A test can be added in which a "c" highlight query inherits from the runtime "c" highlight query, but that is effectively the same as the extends test. I can add that in if desired though.

wookayin commented on test/functional/treesitter/highlight_spec.lua:
```diff
@@ -523,6 +523,37 @@ describe('treesitter highlighting (C)', function()
+    ]])
+
+    exec_lua [[
+      vim.treesitter.query.set("c", "highlights", "[\";\"] @variable", { extends = true })
```
You could use `'` to avoid escaping.

wookayin commented on test/functional/treesitter/highlight_spec.lua:
```diff
@@ -523,6 +523,37 @@ describe('treesitter highlighting (C)', function()
+        [5] = { foreground = Screen.colors.SlateBlue },
+      },
+    }
+  end)
```
In addition, the behavior `vim.treesitter.query.get()` after `vim.treesitter.query.set()` can also be tested.

justinmk commented:
The PR template is not optional. It asks for a problem/solution explanation. Linking to an issue is not sufficient. 

Same for the commit message.

tadmccorkle commented:
> The PR template is not optional. It asks for a problem/solution explanation. Linking to an issue is not sufficient.
> 
> Same for the commit message.

Sorry about that. Description updated and most recent commit includes similar info.

tadmccorkle commented on runtime/lua/vim/treesitter/query.lua:
```diff
@@ -208,16 +208,54 @@ function M.set_query(...)
+    if opts.extends then
+      local files = M.get_files(lang, query_name)
+      vim.list_extend(query_files, files)
+    end
```
Just to clarify, you're referring to multiple extensions being applied by calling `set` multiple times, correct? So multiple plugins can extend the same query? If so, reading runtime files and parsing can be delayed until `vim.treesitter.query.get` is called instead of within `set`.

I noticed one other thing while adding the test for `get` after `set`. `vim.treesitter.query.get` is now memoized, so a call to `get` isn't guaranteed to include the most recent `set` if `get` was previously called. Should the memoize cache be invalidated for a given query after a call to `set`?

lewis6991 commented:
Is there an example use case that is motivating this PR?

clason commented:
Web templates, or expensive injections that we omit, or customizing conceal in markdown.

lewis6991 commented:
And that can't be achieved using `;extends` in query files? 

A fleshed out example would help inform decisions as I'm not convinced of the solution here.

wookayin commented:
I think #26965 would be a more suitable place for the use case discussion whereas this PR can more focus on the technical (implementation) side, but one use case I once needed this for is *dynamic* injections: For example, on python docstrings, one could use ReST, markdown, numpy-style, or any other styles to inject which can be determined only dynamically (depending on the buffer contents itself or per-project, user-level config, etc.) rather than static query files with `; extends`.

lewis6991 commented:
Thanks, that's a useful use case.

 I would expect dynamic queries like that to be per-buffer as opposed to per-lang, so that's something this PR would need to amend.

tadmccorkle commented:
@wookayin Is this extension behavior more of what you had in mind? It allows multiple callers to extend the same query and gives the option to set/extend queries per buffer or per lang. Still have some issues with memoizing though (refer to the most recent commit message).

Converting this back to a draft as I haven't had time to thoroughly test this or implement a consistent solution to invalidate the memoize cache.

tadmccorkle committed:
feat(func): add memoize function cache invalidation

Updates memoize function to return a second function that either:

- Invalidates the entire memoize cache when provided no arguments, or
- Removes a specific function call from the cache when provided that
  function call's arguments

tadmccorkle commented:
Regarding memoize function cache invalidation, it felt odd to change set a query via `vim.treesitter.query.set` and not have the changes reflected until a GC occurred sometime later. That behavior may not be important enough (or my cache clearing implementation may need more work/thought) to warrant including that here.

lewis6991 commented on runtime/lua/vim/func/_memoize.lua:
```diff
@@ -33,27 +33,54 @@ local function resolve_hash(hash)
 --- @return F
+--- @overload fun(hash: vim.func.Hash, fn: fun(...): any): vim.func.MemoizeTable
 return function(hash, fn)
   vim.validate({
```
A better way of doing this would.be for `_memoize` to return 2 functions instead of one. That way you are not polluting the original function.

MariaSolOs commented:
@tadmccorkle are you still working on this? If not, I would like to continue your work :)

tadmccorkle commented:
> @tadmccorkle are you still working on this? If not, I would like to continue your work :)

@MariaSolOs This slipped my mind not too long after my last commit, so I'm not actively working on it. I saw you opened a PR to continue it - thanks! Sorry for letting it sit out so long.

MariaSolOs commented:
> @MariaSolOs This slipped my mind not too long after my last commit, so I'm not actively working on it. I saw you opened a PR to continue it - thanks! Sorry for letting it sit out so long.

No worries at all! A lot of the caching logic will be part of a separate PR so I found it easier to open a new one, yet I will base the implementation on what you've done so far. Thanks to you!

justinmk commented:
https://github.com/neovim/neovim/pull/30257

clason commented:
Noting here for the sake of the record: I think this PR tried to do too much at once (buffer-local queries, refactoring modules to make use of that). Instead, we should first focus _only_ on letting `query.set` extend queries; for that it's enough to ("just") rewrite the query logic so that `set` queries are treated as on top of the "stack" (i.e., are always considered before queries on `runtimepath`) instead of short-circuiting the `rtp` query logic. This way, it's enough to add `; extends` to the query, same as on `rtp`. (We'd also get `; inherits` for free.)
